### PR TITLE
HBASE-22370 ByteBuf LEAK ERROR

### PR DESCRIPTION
**Pull request by binlijin**
If FanOutOneBlockAsyncDFSOutput#endBlock throw Exception before call "buf.release();", this buf has not chance to release.
In CallRunner if the call skipped or Dropping timed out call, the call do not call cleanup.

### CODE REVIEW

File: hbase-server/src/main/java/org/apache/hadoop/hbase/ipc/ServerCall.java [hunk 1]
```diff
@@ -166,7 +166,7 @@ private void release(int mask) {
   }
 
   @Override
-  public void cleanup() {
+  public synchronized void cleanup() {
```
Comment (wchevreuil) [inline hbase-server/src/main/java/org/apache/hadoop/hbase/ipc/ServerCall.java]: Does this really requires to be _synchronized_? If so, you might also want to synchronize _releaseByWAL_ method?

Comment (wchevreuil) [inline hbase-server/src/main/java/org/apache/hadoop/hbase/ipc/ServerCall.java]: Does this really requires to be synchronized? If so, you might also want to synchronize releaseByWAL method?

**Comment by Apache-HBase** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| :blue_heart: |  reexec  |   1m 12s |  Docker mode activated.  |
||| _ Prechecks _ |
| :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| :yellow_heart: |  test4tests  |   0m  0s |  The patch doesn't appear to include any new or modified tests. Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch.  |
||| _ master Compile Tests _ |
| :green_heart: |  mvninstall  |   6m 52s |  master passed  |
| :green_heart: |  compile  |   0m 58s |  master passed  |
| :green_heart: |  checkstyle  |   1m 32s |  master passed  |
| :green_heart: |  shadedjars  |   5m  1s |  branch has no errors when building our shaded downstream artifacts.  |
| :green_heart: |  javadoc  |   0m 39s |  master passed  |
| :blue_heart: |  spotbugs  |   4m 40s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| :green_heart: |  findbugs  |   4m 38s |  master passed  |
||| _ Patch Compile Tests _ |
| :green_heart: |  mvninstall  |   5m 29s |  the patch passed  |
| :green_heart: |  compile  |   0m 57s |  the patch passed  |
| :green_heart: |  javac  |   0m 57s |  the patch passed  |
| :green_heart: |  checkstyle  |   1m 25s |  the patch passed  |
| :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| :green_heart: |  shadedjars  |   5m  1s |  patch has no errors when building our shaded downstream artifacts.  |
| :green_heart: |  hadoopcheck  |  17m 15s |  Patch does not cause any errors with Hadoop 2.8.5 2.9.2 or 3.1.2.  |
| :green_heart: |  javadoc  |   0m 35s |  the patch passed  |
| :green_heart: |  findbugs  |   4m 34s |  the patch passed  |
||| _ Other Tests _ |
| :broken_heart: |  unit  | 264m  7s |  hbase-server in the patch failed.  |
| :green_heart: |  asflicense  |   0m 28s |  The patch does not generate ASF License warnings.  |
|  |   | 327m 33s |   |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hbase.security.access.TestZKPermissionWatcher |
|   | hadoop.hbase.master.TestMasterShutdown |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.3 Server=19.03.3 base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-720/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/720 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs findbugs shadedjars hadoopcheck hbaseanti checkstyle compile |
| uname | Linux 217fb3dca917 4.15.0-58-generic #64-Ubuntu SMP Tue Aug 6 11:12:41 UTC 2019 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /home/jenkins/jenkins-slave/workspace/HBase-PreCommit-GitHub-PR_PR-720/out/precommit/personality/provided.sh |
| git revision | master / 9e628b715d |
| Default Java | 1.8.0_181 |
| unit | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-720/1/artifact/out/patch-unit-hbase-server.txt |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-720/1/testReport/ |
| Max. process+thread count | 5270 (vs. ulimit of 10000) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-720/1/console |
| versions | git=2.11.0 maven=2018-06-17T18:33:14Z) findbugs=3.1.11 |
| Powered by | Apache Yetus 0.11.0 https://yetus.apache.org |


This message was automatically generated.

File: hbase-server/src/main/java/org/apache/hadoop/hbase/ipc/ServerCall.java (hunk 1 above)
Comment (binlijin) [reply hbase-server/src/main/java/org/apache/hadoop/hbase/ipc/ServerCall.java]: This method can be called by Handler or Netty Worker concurrently.   
But releaseByWAL do not.

Comment (Apache9) [reply hbase-server/src/main/java/org/apache/hadoop/hbase/ipc/ServerCall.java]: But I think the implementation is thread safe?

File: hbase-server/src/main/java/org/apache/hadoop/hbase/ipc/CallRunner.java [hunk 2]
```diff
@@ -88,6 +88,7 @@ public void setStatus(MonitoredRPCHandler status) {
    * Cleanup after ourselves... let go of references.
    */
   private void cleanup() {
+    this.call.cleanup();
```
Comment (Apache9) [inline hbase-server/src/main/java/org/apache/hadoop/hbase/ipc/CallRunner.java]: So this is for safety. And I checked the code, drop is only called in AdaptiveLifoCoDelCallQueue, maybe this is why we do not catch this problem? Can we add a UT for this?

File: hbase-server/src/main/java/org/apache/hadoop/hbase/io/asyncfs/FanOutOneBlockAsyncDFSOutput.java [hunk 3]
```diff
@@ -550,6 +550,10 @@ private void endBlock() throws IOException {
    */
   @Override
   public void recoverAndClose(CancelableProgressable reporter) throws IOException {
+    if (buf != null) {
```
Comment (Apache9) [inline hbase-server/src/main/java/org/apache/hadoop/hbase/io/asyncfs/FanOutOneBlockAsyncDFSOutput.java]: So we missed this. Nice catch.

**Comment by Apache-HBase** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| :blue_heart: |  reexec  |   1m  9s |  Docker mode activated.  |
||| _ Prechecks _ |
| :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 1 new or modified test files.  |
||| _ master Compile Tests _ |
| :green_heart: |  mvninstall  |   6m  3s |  master passed  |
| :green_heart: |  compile  |   0m 57s |  master passed  |
| :green_heart: |  checkstyle  |   1m 31s |  master passed  |
| :green_heart: |  shadedjars  |   5m  1s |  branch has no errors when building our shaded downstream artifacts.  |
| :green_heart: |  javadoc  |   0m 36s |  master passed  |
| :blue_heart: |  spotbugs  |   4m 29s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| :green_heart: |  findbugs  |   4m 27s |  master passed  |
||| _ Patch Compile Tests _ |
| :green_heart: |  mvninstall  |   5m 31s |  the patch passed  |
| :green_heart: |  compile  |   0m 57s |  the patch passed  |
| :green_heart: |  javac  |   0m 57s |  the patch passed  |
| :green_heart: |  checkstyle  |   1m 28s |  the patch passed  |
| :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| :green_heart: |  shadedjars  |   4m 59s |  patch has no errors when building our shaded downstream artifacts.  |
| :green_heart: |  hadoopcheck  |  17m 26s |  Patch does not cause any errors with Hadoop 2.8.5 2.9.2 or 3.1.2.  |
| :green_heart: |  javadoc  |   0m 35s |  the patch passed  |
| :green_heart: |  findbugs  |   4m 47s |  the patch passed  |
||| _ Other Tests _ |
| :broken_heart: |  unit  | 249m  5s |  hbase-server in the patch failed.  |
| :green_heart: |  asflicense  |   0m 28s |  The patch does not generate ASF License warnings.  |
|  |   | 311m 56s |   |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hbase.security.access.TestSnapshotScannerHDFSAclController |
|   | hadoop.hbase.client.TestAsyncTableAdminApi3 |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.3 Server=19.03.3 base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-720/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/720 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs findbugs shadedjars hadoopcheck hbaseanti checkstyle compile |
| uname | Linux 13390d5e5df2 4.15.0-58-generic #64-Ubuntu SMP Tue Aug 6 11:12:41 UTC 2019 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /home/jenkins/jenkins-slave/workspace/HBase-PreCommit-GitHub-PR_PR-720/out/precommit/personality/provided.sh |
| git revision | master / 7924ba39e7 |
| Default Java | 1.8.0_181 |
| unit | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-720/2/artifact/out/patch-unit-hbase-server.txt |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-720/2/testReport/ |
| Max. process+thread count | 5293 (vs. ulimit of 10000) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-720/2/console |
| versions | git=2.11.0 maven=2018-06-17T18:33:14Z) findbugs=3.1.11 |
| Powered by | Apache Yetus 0.11.0 https://yetus.apache.org |


This message was automatically generated.

File: hbase-server/src/main/java/org/apache/hadoop/hbase/ipc/ServerCall.java (hunk 1 above)
Comment (binlijin) [reply hbase-server/src/main/java/org/apache/hadoop/hbase/ipc/ServerCall.java]: OK, i make a mistake, it is thread safe now.

File: hbase-server/src/main/java/org/apache/hadoop/hbase/ipc/CallRunner.java (hunk 2 above)
Comment (binlijin) [reply hbase-server/src/main/java/org/apache/hadoop/hbase/ipc/CallRunner.java]: if (call.disconnectSince() >= 0) {
        if (RpcServer.LOG.isDebugEnabled()) {
          RpcServer.LOG.debug(Thread.currentThread().getName() + ": skipped " + call);
        }
        return;
      }
      call.setStartTime(System.currentTimeMillis());
      if (call.getStartTime() > call.getDeadline()) {
        RpcServer.LOG.warn("Dropping timed out call: " + call);
        return;
      }

When this two case, the call.cleanup() also not called before.

Comment (Apache9) [reply hbase-server/src/main/java/org/apache/hadoop/hbase/ipc/CallRunner.java]: Yes, I got them after seeing your UT.

**Comment by Apache-HBase** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| :blue_heart: |  reexec  |   1m  9s |  Docker mode activated.  |
||| _ Prechecks _ |
| :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 1 new or modified test files.  |
||| _ master Compile Tests _ |
| :green_heart: |  mvninstall  |   5m 57s |  master passed  |
| :green_heart: |  compile  |   0m 58s |  master passed  |
| :green_heart: |  checkstyle  |   1m 31s |  master passed  |
| :green_heart: |  shadedjars  |   5m  2s |  branch has no errors when building our shaded downstream artifacts.  |
| :green_heart: |  javadoc  |   0m 37s |  master passed  |
| :blue_heart: |  spotbugs  |   4m 27s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| :green_heart: |  findbugs  |   4m 24s |  master passed  |
||| _ Patch Compile Tests _ |
| :green_heart: |  mvninstall  |   5m 42s |  the patch passed  |
| :green_heart: |  compile  |   0m 58s |  the patch passed  |
| :green_heart: |  javac  |   0m 58s |  the patch passed  |
| :green_heart: |  checkstyle  |   1m 32s |  the patch passed  |
| :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| :green_heart: |  shadedjars  |   5m  1s |  patch has no errors when building our shaded downstream artifacts.  |
| :green_heart: |  hadoopcheck  |  17m 26s |  Patch does not cause any errors with Hadoop 2.8.5 2.9.2 or 3.1.2.  |
| :green_heart: |  javadoc  |   0m 34s |  the patch passed  |
| :green_heart: |  findbugs  |   4m 38s |  the patch passed  |
||| _ Other Tests _ |
| :green_heart: |  unit  | 241m 12s |  hbase-server in the patch passed.  |
| :green_heart: |  asflicense  |   0m 28s |  The patch does not generate ASF License warnings.  |
|  |   | 303m 45s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.3 Server=19.03.3 base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-720/3/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/720 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs findbugs shadedjars hadoopcheck hbaseanti checkstyle compile |
| uname | Linux 919a22e4bb24 4.15.0-58-generic #64-Ubuntu SMP Tue Aug 6 11:12:41 UTC 2019 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /home/jenkins/jenkins-slave/workspace/HBase-PreCommit-GitHub-PR_PR-720/out/precommit/personality/provided.sh |
| git revision | master / 7924ba39e7 |
| Default Java | 1.8.0_181 |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-720/3/testReport/ |
| Max. process+thread count | 4644 (vs. ulimit of 10000) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-720/3/console |
| versions | git=2.11.0 maven=2018-06-17T18:33:14Z) findbugs=3.1.11 |
| Powered by | Apache Yetus 0.11.0 https://yetus.apache.org |


This message was automatically generated.

**Pull request merged**